### PR TITLE
Rewrite prune_old_outbound_device_pokes for efficiency

### PR DESCRIPTION
**Pull request by richvdh**
* make sure we clear out all but one update for the user
 * ~~there is no need to delete from `device_lists_outbound_last_success`: sticking with the last successful stream id is fine.~~

Based on #7157.

### CODE REVIEW

File: tests/federation/test_federation_sender.py [hunk 1]
```diff
@@ -370,6 +370,43 @@ def test_unreachable_server(self):
...
+        mock_send_txn.side_effect = self.record_transaction
+        self.hs.get_federation_sender().send_device_messages("host2")
+        self.pump()
+
+        # there should be a single update for this user.
+        self.assertEqual(len(self.edus), 1)
+        edu = self.edus.pop(0)
+        self.assertEqual(edu["edu_type"], "m.device_list_update")
+        c = edu["content"]
+        self.assertEqual(c["prev_id"], [])
```
Comment (erikjohnston) [inline tests/federation/test_federation_sender.py]: Can we get a test where we successfully send an update before we prune please? Now that we're not deleting from `device_lists_outbound_last_success` its not clear what the `prev_id` will be in that case.

Separately, the spec doesn't actually say that the remote should resync when `prev_id` is empty, though Synapse does. We should either a) fix Synapse to point to a dropped update or b) update the spec. I guess I'm happy with not trying to fix it in this PR, but if so we should at least file an issue or something

File: tests/federation/test_federation_sender.py [hunk 2]
```diff
@@ -147,3 +153,331 @@ def test_send_receipts_with_backoff(self):
...
+            e2e_handler.upload_signatures_for_device_keys(
+                u1, {u1: {"D1": d1_json, "D2": d2_json}},
+            )
+        )
+        self.assertEqual(ret["failures"], {})
+
+        # expect two edus, in one or two transactions. We don't know what order the
+        # devices will be updated.
+        self.assertEqual(len(self.edus), 2)
+        stream_id = None  # FIXME: there is a discontinuity in the stream IDs: see #7142
```
Comment (erikjohnston) [inline tests/federation/test_federation_sender.py]: Can we fix this?

Comment (richvdh) [reply tests/federation/test_federation_sender.py]: I guess so, though I think it's an orthogonal problem, and it's not clear to me that it's a priority?

File: tests/federation/test_federation_sender.py (hunk 1 above)
Comment (richvdh) [reply tests/federation/test_federation_sender.py]: Yeah, thanks for making me think about this properly. Removing the deletion from `device_lists_outbound_last_success` was the wrong thing.

Fixing it isn't entirely trivial: have raised #7173 to track

File: tests/federation/test_federation_sender.py (hunk 2 above)
Comment (erikjohnston) [reply tests/federation/test_federation_sender.py]: Oh, sorry. For some reason I thought that this PR would mean we could remove that. Nevermind.

**Pull request merged**